src/main/java/org/radarbase/kafka/connect/transforms/KeyValueTransform.java

```diff
@@ -0,0 +1,131 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.radarbase.kafka.connect.transforms;
+
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.connect.connector.ConnectRecord;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.transforms.Transformation;
+import static org.apache.kafka.connect.transforms.util.Requirements.requireMap;
+import static org.apache.kafka.connect.transforms.util.Requirements.requireStruct;
+
+import java.util.*;
+
+/**
+ *
+ *
+ * This transforms records by copying the record key into the record value.
+ */
+public class KeyValueTransform<R extends ConnectRecord<R>> implements Transformation<R> {
+  private static final String PURPOSE = "copying fields from key to value";
+  private static final String TIMESTAMP_FIELD = "timestamp";
+  private static final String KEYVALUE_SCHEMA_NAME = "KeyToValue";
+  private static final Set<String> TIME_FIELDS = new HashSet<>(Arrays.asList("time", "timeReceived", "timeCompleted"));
+
+  @Override
+  public R apply(R r) {
+    if (r.valueSchema() == null) {
+      return applySchemaless(r);
+    } else {
+      return applyWithSchema(r);
+    }
+  }
+
+  private R applyWithSchema(R r) {
+    SchemaBuilder schemaBuilder = SchemaBuilder.struct().name(KEYVALUE_SCHEMA_NAME);
+    schemaBuilder = updateSchema(schemaBuilder, r.keySchema());
+    schemaBuilder = updateSchema(schemaBuilder, r.valueSchema());
+    schemaBuilder.field(TIMESTAMP_FIELD, Schema.INT64_SCHEMA);
+    Schema schema = schemaBuilder.build();
+
+    final Struct recordValue = requireStruct(r.value(), PURPOSE);
+    final Struct recordKey = requireStruct(r.key(), PURPOSE);
+    Struct newData = new Struct(schema);
+    newData = addValuesToNewSchema(newData, recordKey, r.keySchema());
+    newData = addValuesToNewSchema(newData, recordValue, r.valueSchema());
+    newData.put(TIMESTAMP_FIELD, r.timestamp());
+
+    return r.newRecord(r.topic(), r.kafkaPartition(), null, null, schema, newData, r.timestamp());
+  }
+
+  private Struct addValuesToNewSchema(Struct newData, Struct oldData, Schema oldSchema){
+    for (Field field: oldSchema.fields()) {
+      String fieldName = field.name();
+      Object fieldVal = oldData.get(field);
+      if(TIME_FIELDS.contains(fieldName))
+        fieldVal = convertTimestamp(fieldVal);
+      newData.put(fieldName, fieldVal);
+    }
+    return newData;
+  }
+
+
+  private SchemaBuilder updateSchema(SchemaBuilder schemaBuilder, Schema oldSchema){
+    for (Field field: oldSchema.fields()) {
+      String fieldName = field.name();
+      if(TIME_FIELDS.contains(fieldName))
```

Comment:
Perhaps add: `&& field.type().schema().type() == DOUBLE)`. I’m writing this from my phone so the exact syntax will be different...

Reply:
Okay thanks